### ISSUE TITLE
General Discussion

### ISSUE THREAD

**Issue by qzed** [linux-surface/surface-pro-x#7]
Issue for general discussions.

**Comment by qzed** [comment]
Feel free to open a new issue for that! Unfortunately, this is something I know practically nothing about.

**Comment by liuyishengalan** [comment]
Thanks for your contribution to making Linux usable on SPX.
I have no experience in making Adreno GPU work in Linux, can you provide more detailed steps to guide us.
Currently, I installed the xf86-video-freedreno package in Arch Linux ARM and face a problem:
Xorg.0.log:
`NO msm DRM/KMS, fallingback to fbdev/kgsl`
GUI can show up without freedreno -> no GPU acceleration.
The kernel was compiled using spx/next-20220607 branch/ and boot up with DT.

**Comment by qzed** [comment]
Can you post a full dmesg log? Also what's the putput of `ls -l /sys/class/drm`?


In terms of user-space software, I only explicitly installed `mesa` (next to `gnome` on Arch, which may have pulled in some other dependencies). I haven't tried with xorg yet (only Wayland), but that shouldn't really make a difference.

**Comment by liuyishengalan** [comment]
There are no files under `/sys/class/drm` directory. My SPX cannot connect to the internet via a USB ethernet port, that's a bit hard to post the log. I will try to use mesa driver to see if anything will work.

**Comment by qzed** [comment]
In that case, I don't think mesa will work either. Does `/sys/bus/platform/devices/2c00000.gpu` exist?

**Comment by liuyishengalan** [comment]
2c00000.gpu does exist.

**Comment by qzed** [comment]
Do you have an SQ1 or SQ2? Also at this point I think we need a dmesg log.

**Comment by liuyishengalan** [comment]
SQ1 [dmesg info](https://gist.github.com/liuyishengalan/8fb84c842db5c7e5af8b45ea43f41b38)

**Comment by qzed** [comment]
Okay, so might be a difference between SQ1 and SQ2. What about `cat /sys/kernel/debug/devices_deferred`?

**Comment by liuyishengalan** [comment]
`/sys/kernel/debug/devices_deferred` is a blank file

**Comment by qzed** [comment]
That means that no devices (that have drivers) wait on other drivers to load. I can't seem to find any reference to pretty much anything GPU related in your dmesg log. Can you post your kernel config?

**Comment by liuyishengalan** [comment]
The config I used is [spx-5.19.rc1-arch.config](https://github.com/qzed/aarch64-kernel-configs/blob/main/full/spx-5.19.rc1-arch.config)

**Comment by qzed** [comment]
Can you try and export the final/actual config with `zcat /proc/config.gz > config` or something like that? There seem to be a bunch of other devices missing as well... at the very least SAM.

**Comment by liuyishengalan** [comment]
https://gist.github.com/liuyishengalan/fb0e8233f844acdcdb2badcdd1c9defd

**Comment by qzed** [comment]
Okay, so the config checks out... it kinda looks like everything with drivers built as modules doesn't work... Can you check if they're accessible (e.g. run `modinfo msm` or something like that).

**Comment by liuyishengalan** [comment]
`modinfo msm` output: 
`modinfo: ERROR: Module msm not found`

**Comment by qzed** [comment]
So I guess it can't find any modules... the partitions are mounted properly? What about `ls -l /usr/lib/modules/<your-kernel-version>`?

**Comment by liuyishengalan** [comment]
```
total 24
-rw-r--r-- 1 root root 45 Jun 12  2022 modules.alias
-rw-r--r-- 1 root root 12 Jun 12  2022 modules.alias.bin
-rw-r--r-- 1 root root  0 Jun 12  2022 modules.builtin.alias.bin
-rw-r--r-- 1 root root  0 Jun 12  2022 modules.builtin.bin
-rw-r--r-- 1 root root  0 Jun 12  2022 modules.dep
-rw-r--r-- 1 root root 12 Jun 12  2022 modules.dep.bin
-rw-r--r-- 1 root root  0 Jun 12  2022 modules.devname
-rw-r--r-- 1 root root 55 Jun 12  2022 modules.softdep
-rw-r--r-- 1 root root 49 Jun 12  2022 modules.symbols
-rw-r--r-- 1 root root 12 Jun 12  2022 modules.symbols.bin
```

**Comment by qzed** [comment]
Okay, there should be a `kernel` directory with modules there... So I guess either something isn't mounted properly or whatever you used for installation missed some parts.

**Comment by liuyishengalan** [comment]
> Okay, there should be a `kernel` directory with modules there... So I guess either something isn't mounted properly or whatever you used for installation missed some parts.

Thanks for the hint. I will look into it.

**Comment by liuyishengalan** [comment]
> 

Solved. `copymods` was mounted under `/lib/modules`, all kernel modules will be deleted after reboot. SAM and GPU works on SQ1

**Comment by chenx-dust** [comment]
I got some problems while booting Fedora 36 installation on my surface pro x.
First, I downloaded fedora image for aarch64 on its website. And then I build linux kernel from [linux-surface/kernel:spx/v5.19-rc8](https://github.com/linux-surface/kernel/tree/spx/v5.19-rc8) with its dtbs, using config `full/spx-5.19.rc1-fedora_36.config` from [qzed/aarch64-kernel-configs](https://github.com/qzed/aarch64-kernel-configs). Afterwards, I replace the original vmlinuz with built file `arch/arm64/boot/Image.gz` and edit grub.cfg. However, I failed to boot after loading initrd.

Here is my `grub.cfg`:
```
set default="1"

function load_video {
  if [ x$feature_all_video_module = xy ]; then
    insmod all_video
  else
    insmod efi_gop
    insmod efi_uga
    insmod ieee1275_fb
    insmod vbe
    insmod vga
    insmod video_bochs
    insmod video_cirrus
  fi
}

load_video
set gfxpayload=keep
insmod gzio
insmod part_gpt
insmod ext2

set timeout=60
### END /etc/grub.d/00_header ###

search --no-floppy --set=root -l 'Fedora-WS-Live-36-1-5'

### BEGIN /etc/grub.d/10_linux ###
menuentry 'Start Fedora-Workstation-Live 36' --class red --class gnu-linux --class gnu --class os {
  echo "Loading DTB..."
  devicetree /images/pxeboot/sc8180x-surface-pro-x.dtb
  echo "Loading kernel.."
  linux /images/pxeboot/vmlinuz clk_ignore_unused root=live:CDLABEL=FEDORA-WS-L rd.live.image quiet rhgb efi=novamap
  echo "Loading initrd..."
  initrd /images/pxeboot/initrd.img
}
menuentry 'Test this media & start Fedora-Workstation-Live 36' --class red --class gnu-linux --class gnu --class os {
  echo "Loading DTB..."
  devicetree /images/pxeboot/sc8180x-surface-pro-x.dtb
  echo "Loading kernel.."
  linux /images/pxeboot/vmlinuz clk_ignore_unused root=live:CDLABEL=FEDORA-WS-L rd.live.image rd.live.check quiet efi=novamap
  echo "Loading initrd..."
  initrd /images/pxeboot/initrd.img
}
submenu 'Troubleshooting -->' {
  menuentry 'Install Fedora-Workstation-Live 36 in basic graphics mode' --class red --class gnu-linux --class gnu --class os {
    echo "Loading DTB..."
    devicetree /images/pxeboot/sc8180x-surface-pro-x.dtb
    echo "Loading kernel.."
    linux /images/pxeboot/vmlinuz clk_ignore_unused root=live:CDLABEL=FEDORA-WS-L rd.live.image nomodeset quiet rhgb efi=novamap
    echo "Loading initrd..."
    initrd /images/pxeboot/initrd.img
  }
}

```

I'm a beginner in linux kernel field, would you please lend me a hand?

**Comment by chenx-dust** [comment]
Em, it seems that I've found a corresponding explanation in https://github.com/linux-surface/grub/issues/1

**Comment by nickdepinet** [comment]
I've recently gotten ahold of a [windows dev kit 2023](https://learn.microsoft.com/en-us/windows/arm/dev-kit/), which appears to be a surface pro 9 (5g) / sq3 in a box, and I'm attempting to boot linux on it with these patches.

Progress so far - 

1) Booting base Ubuntu 22.04.1 LTS arm image - Hangs at:
```
Booting a command list


Synchronous Exception at 0x0000000C1B17927C
```

2) Building a "default" profile image with the [image tool](https://github.com/linux-surface/aarch64-arch-mkimg) - Hangs at:
```
Loading initramfs...
Booting into Surface kernel...
EFI stub: Booting Linux Kernel...
EFI stub: Using DTB from configuration table
EFI stub: Exiting boot services...
```

Barring better suggestions, I'm going to try to extract the firmware with the script from aarch64-firmware and stuff it into the arch image above and see where that gets booting

**Comment by qzed** [comment]
1) could possibly be fixed by https://github.com/linux-surface/grub/commit/55b95a388414a3d850e24fd2ae8fac100ffd9319. So you could try to use e.g. the pre-built grub from https://github.com/linux-surface/grub-image-aarch64/releases/tag/fedora-37-2 with your Ubuntu kernel. Alternatively, just use the image from 2) which contains the patched grub, but remove the `devicetree` option from the boot entry. That should get you booting in ACPI mode.

2) is way before any firmware gets loaded, so I doubt that that will make any difference. Can you make sure that `efi=novamap` is set (press `e` in the grub menu to edit)? Also, this is booting with the SPX DT, and I'm not sure if that's a good idea. So as I've said above, you can remove the `devicetree` option in the grub config and it should hopefully get you a bit further.

You can also experiment with `earlycon=efifb`.

In the long run, you'll probably want to have a look at the DT of the sc8280xp (that is 8cx Gen 3; it's also used in the Thinkpad X13s) and use that as a base. Also, you might want to join the `oftc/#aarch64-laptops` IRC. There are a lot more experienced people who also might have some ideas.

**Comment by ParticleG** [comment]
Really looking forward to this project. Any new developments currently？

**Comment by qzed** [comment]
For the dev-kit see #43. As far as I understand, it's booting via ACPI with display support, but a DT will likely be required to get the GPU working.

For the SPX: I'm a bit busy right now, so progress is somewhat slow. I'm working on getting some patches for UEFI variables (that are also useful for other Qualcomm devices) upstream. I'll also need to revisit the GRUB patches (completely forgot about those). So hopefully I can find the time to work on that over easter.

There's also https://github.com/linux-surface/aarch64-arch-mkimg/pull/5  to build a DT-bootable disk image that should hopefully make installation a bit easier. Unfortunately, that seems to building somewhat randomly (fails downloading a package from our repo) and I'm not exactly sure why.

**Comment by TimoRams** [comment]
Hey, I think it's really great to see progress like the arm64 releases, but is there any progress on the audio or webcam drivers?

**Comment by qzed** [comment]
@TimoRams Not on the SPX side, unfortunately.

I'm not sure how far along webcam support is on the X13s (that could probably at least partially be adapted for the ARM-based SP9 and DevKit). AFAIK audio should work on that though (but again might have to be adapted in some parts, at least integrated in some DTS).

**Comment by denysvitali** [comment]
Does anyone know if there is an issue with acceleration / graphics? It seems like when there is high CPU usage, the device has a very noticeable lag (to the point that actions are delayed by seconds).

This happens on Wayland (GNOME and Sway), but AFAIK HW Acceleration is enabled

**Comment by jglathe** [comment]
Hmm. Can't say that, really, I'm using the WDK and Thinkpad X13s on Windows 11 and Linux, and on Linux it's not perfect, but at least on par with Windows 11. More like a lttle faster. There are some USB-related issues, meaning if there's a lot of traffic on it, you would see lags, but as daily driver... nope What#s your setup?

**Comment by denysvitali** [comment]
This mostly happens when there is really high CPU usage such as `pacman` decompressing with `xz` some packages.
Of course this is "expected to happen" when the load is high, but I never faced such an issue outside of my SPX under high load - which makes me think that the GPU is doing SW rendering (?).

**Comment by TimoRams** [comment]
Can someone explain to me how to install the arch linux from the aarch64-arch-mkimg on the nvme since the instructions on how to install it is different from x86. The other question would be how to copy the surface kernal to the manjaro install usb for example. I'm pretty new to all this but I would like to run Arch/Manjaro on my surface.

**Comment by xmod17** [comment]
anyone try to boot android, i try but never success

**Comment by matterhorn103** [comment]
Hi everyone! I have an SQ1 Pro X and am getting increasingly frustrated with successive Windows 11 updates degrading the performance, and turning what was, on release, really a very nice device, into a bit of a pain in the arse.

I'm also much more keen to be running Linux on everything these days than I was when I bought it. So I'd love to put it on my SPX but I'm a little unclear as to how exactly to go about it.

Is the tracking issue representative of the current state i.e. things basically work except sound, sensors, webcams? "External Display Ports" don't work, but what about HDMI alt-mode over USB-C?

If I am to install it, how much of the support has been upstreamed? Can I try a normal openSUSE Tumbleweed installer these days, like I did once with an old SP4? (Would be my preferred distro.) Or do I need to use this special Arch image?

I use Linux on my main desktop PC but have essentially zero experience with the kernel and almost no knowledge of its inner workings. However, I do actually have one, and it's not my primary device, so I am very happy to try out things and investigate things, send logs etc. or anything that will help further development. Writing anything myself is probably beyond me (basically only know Python), but maybe I can still assist in other ways, if guided? :)

**Comment by qzed** [comment]
> but what about HDMI alt-mode over USB-C

I haven't really tested that.

> If I am to install it, how much of the support has been upstreamed? Can I try a normal openSUSE Tumbleweed installer these days, like I did once with an old SP4? (Would be my preferred distro.) Or do I need to use this special Arch image?

The critical parts are still not upstream, most notably the device tree. Even then it would probably not be straightforward to install, because Linux doesn't support ACPI PEPs, with Qualcomm makes heavy use of with those kind of devices. I think the easiest we can do is create some kind of install media like I've done [here](https://github.com/linux-surface/aarch64-arch-mkimg). But note that that's not really user-friendly.

**Comment by matterhorn103** [comment]
Ah, that does seem like a quite a lot of hoops still yeah...

**Comment by kikispcmr** [comment]
I'm facing some issues with getting the surface pro x to actually boot into the installed linux image (booting into the patched kernel with the patched firmware installed). 

The exact errors displayed on boot are:
`/sys/firmware/efi/efivars: mount(2) system call failed: operation not supported`
Where dmseg returns:
`ERROR device 'UUID=...' not found. Skipping fsck.`
Where after I'm just dropped in the emergency shell where I can't do much. 

I've verified everything that I could.  The UUIDs are correct and consistent across fstab and the final grub config. I've regenerated the GRUB config, regenerated initramfs and also ensured secure boot is turned off. Efivars are also mounted, and present, so I'm doubtful the system is getting hang up on that. I've tried the installation many times with slight variations hoping that the device tree gets properly loaded up, but nothing has worked so far. Any suggestions for further troubleshooting?

**Comment by superxkooda** [comment]
Set this up yesterday, finished setting up hyprland today. This rocks man! A few things still not working but this is to be expected with what is ultimately a niche device to want to have Linux on. Never mind the the hoops to get everything that is working working! Thank you, this breaths new life into my little machine. 

The last thing to do to make this perfect is to figure out how to get iptsd to start via a udev rule. I applied the one from the repo, along with the unit file. Udev rules are a bit of a mystery for me, I tend to get them working and it isn't an issue until I have to setup a new system. I have to be missing something silly. 

```
cat /etc/udev/rules.d/50-iptsd.rules 
SUBSYSTEM=="hidraw", ACTION=="add", TAG+="systemd", \
	PROGRAM="/usr/bin/iptsd-check-device --quiet $env{DEVNAME}", \
	PROGRAM="/bin/systemd-escape --path $env{DEVNAME}", \
	ENV{SYSTEMD_WANTS}+="iptsd@$result.service"
```
and the unit file looks like the following:
```
cat /etc/systemd/system/iptsd\@.service 
[Unit]
Description=Intel Precise Touch & Stylus Daemon
Documentation=https://github.com/linux-surface/iptsd
StopWhenUnneeded=yes
BindsTo=%i.device

[Service]
Type=simple
ExecStart=@bindir@/iptsd /%I
```

I am able to get this to run manually by simply using this run line `sudo iptsd $( sudo iptsd-find-hidraw)` and it works as you would expect. Alas the udev rules do not run, and I am not entirely sure why just yet.

If you have an idea, thanks! If you need more info, no problem. If you are unable to help, also ok, I will update if I figure it out.

**Comment by qzed** [comment]
Does running `iptsd-check-device` manually work? I guess if `iptsd-find-hidraw` works it should...

**Comment by superxkooda** [comment]
`iptsd-find-hidraw` appears to be depreciated.
```
$ sudo iptsd-find-hidraw
[10:16:55.752] [warning] iptsd-find-hidraw is deprecated, please use iptsd-foreach
/dev/hidraw0
```
To the udev rule.
`iptsd-check-device` exits with a 0 if the device is the hidraw we are after and a 1 if not. The udev rule appears to be checking every single hidraw${N} in /dev/ and should only attempting to start the service for a good device. 
eg via journalctl
```
hidraw4: Process '/usr/bin/iptsd-check-device --quiet /dev/hidraw4' failed with exit code 1.
```
The provided udev rule appears to be trying and failing to find the device before launching hyprland. I do not have it booting into a graphical login. 
```
hidraw0: Process '/usr/bin/iptsd-check-device --quiet /dev/hidraw0' failed with exit code 1.
```
If ran manually after starting a wayland session, in this case hyprland, it looks like it should have worked just fine.
```
$ sudo /usr/bin/iptsd-check-device --quiet /dev/hidraw0
$ echo $?
0
```
without the quiet option
```
/dev/hidraw0 is an IPTS device!
$ sudo /usr/bin/iptsd-check-device  /dev/hidraw0
[10:24:01.548] [info] Loading config /etc/iptsd.conf.
[10:24:01.548] [info] Connected to device 045E:0C21
[10:24:01.548] [info] Running in Touchscreen mode
[10:24:01.548] [info] /dev/hidraw0 is an IPTS device!

# Another hidraw device for good measure.

$ sudo /usr/bin/iptsd-check-device  /dev/hidraw4
[10:24:04.209] [error] ipts: /dev/hidraw4 is not an IPTS device!
```
I expect the issue may be with when the rule is triggered, because the command will exit with an error if I try it from a tty. Going to play with attempt to adjust how it is triggered, and see if it can be triggered when wayland starts.

On a side note about the unit file. I have tried manually running the unit file but was finding it would kill the process immediately. I had started to remove parts to see if any particular setting was an issue. Removing the BindTo directive in the unit file seems to keep it running. Not sure why yet as the device exists. The %i is the escaped device path eg dev-hidraw0 in this case, and my understanding is that should be correct.

**Comment by qzed** [comment]
> The provided udev rule appears to be trying and failing to find the device before launching hyprland. I do not have it booting into a graphical login.

Can you try removing `--quiet` option? Any chance it could be a permission error (e.g., udev not running iptsd-check-device as root)?

**Comment by superxkooda** [comment]
I am an idiot, the rule was never triggering unless I manually ran `udevadm trigger` manually, and had mistaken the few instances in journalctl's output as had happened during boot.

Will have to tackle the triggering issue later. For now going to trigger via udevadm.

To your point, good idea. I cannot output to standard out because the output is used as the value for $result. Redirecting the output to a file is the way here. I modified the udev rule to be the following
```
SUBSYSTEM=="hidraw", ACTION=="add", TAG+="systemd", \
	PROGRAM="/bin/bash -c 'echo $env{DEVNAME} >> /tmp/iptsd.rule.out; /usr/bin/iptsd-check-device $env{DEVNAME} 2>&1 >> /tmp/iptsd.rule.out' && exit 0 || exit 1", \
	PROGRAM="/bin/systemd-escape --path $env{DEVNAME}", \
	ENV{SYSTEMD_WANTS}+="iptsd@$result.service"
```
The output in /tmp/iptsd.rule.out
```
/dev/hidraw0
[20:12:17.358] [error] core: linux: IOCTL 3229108231 failed: No such device
/dev/hidraw2
/dev/hidraw1
[23:59:26.861] [error] ipts: /dev/hidraw2 is not an IPTS device!
[23:59:26.861] [error] ipts: /dev/hidraw1 is not an IPTS device!
/dev/hidraw3
[23:59:26.872] [error] ipts: /dev/hidraw3 is not an IPTS device!
/dev/hidraw4
[23:59:26.971] [error] ipts: /dev/hidraw4 is not an IPTS device!
/dev/hidraw4
[23:59:27.024] [error] ipts: /dev/hidraw4 is not an IPTS device!
```
Interesting result, it appears hidraw0 is not found at all when ran by a udev rule. Has to be a funky permission issue. Ultimately also may be why the job is not triggering. Going to shove a whoami into that and see if it isn't root tomorrow.

Thanks again, I think you have me pointed in the correct direction here.

:edit: The rules are running as root, guess it is not a permissions issue. Will continue to dig.

**Comment by qzed** [comment]
Hmm, if triggering it manually late after boot still doesn't work then it shouldn't be a driver initialization issue. Maybe you can ask on the IPTSd repo directly, @StollD might have a better idea about what could be going wrong.

**Comment by superxkooda** [comment]
Thank you, I will to do that. Much appreciated.

**Comment by almenscorner** [comment]
Hi, this is fantastic! I managed to get arch running smoothly on my SPX but there is one thing I just cannot get working which is Bluetooth. I've followed the wiki and have all firmware installed. Using `bluetoothctl` I'm not able to see the controller. Any pointers as to what else I need to do?

**Comment by denysvitali** [comment]
> Hi, this is fantastic! I managed to get arch running smoothly on my SPX but there is one thing I just cannot get working which is Bluetooth. I've followed the wiki and have all firmware installed. Using `bluetoothctl` I'm not able to see the controller. Any pointers as to what else I need to do?


See https://github.com/linux-surface/surface-pro-x/issues/47

**Comment by jglathe** [comment]
Hmm on sc8280xp I have a default MAC address set by the dt to avoid issues bringing up the BT host controller.

```
&uart2 {
	pinctrl-0 = <&uart2_default>;
	pinctrl-names = "default";

	status = "okay";

	bluetooth {
		compatible = "qcom,wcn6855-bt";

		vddrfacmn-supply = <&vreg_pmu_rfa_cmn_0p8>;
		vddaon-supply = <&vreg_pmu_aon_0p8>;
		vddwlcx-supply = <&vreg_pmu_wlcx_0p8>;
		vddwlmx-supply = <&vreg_pmu_wlmx_0p8>;
		vddbtcmx-supply = <&vreg_pmu_btcmx_0p8>;
		vddrfa0p8-supply = <&vreg_pmu_rfa_0p8>;
		vddrfa1p2-supply = <&vreg_pmu_rfa_1p2>;
		vddrfa1p8-supply = <&vreg_pmu_rfa_1p7>;

		max-speed = <3200000>;

		local-bd-address = [ 8d fd f0 00 5a ad ];
	};
}
```
Maybe something similar?

**Comment by denysvitali** [comment]
Yes, it's the same issue

**Comment by almenscorner** [comment]
I just reinstalled firmware and it popped up in bluez without issues. Maybe I poked around to much

**Comment by gibfesta** [comment]
hi there, after a long and frustrating month i managed to install arch onto my spq1! 

however, i am having issues with the sound and bluetooth. i get dummy output for sound and for bluetooth even though i see it i cant get it to connect. 

it also appears as though i have problems with the graphics card as with some apps i get the scrambled colours.

any thoughts on this?

thanks

**Comment by McGeoMetro0045** [comment]
Hello there.. is it at all possible for someone to give me an image they made or build me a working image to use.. while arch is not my first choice i would very much enjoy using linux on my device with the current state of windows and cannot find any images on the internet, i do not wish to build my own if possible
